SerialPrograms/Source/CommonFramework/Audio/AudioDisplayWidget.cpp

```diff
@@ -0,0 +1,564 @@
+        return ((float)*reinterpret_cast<const qint8 *>(ptr))/(float)std::numeric_limits<qint8>::max();
+    } else if (format.sampleSize() == 16 && format.sampleType() == QAudioFormat::UnSignedInt) {
+        if (format.byteOrder() == QAudioFormat::LittleEndian)
+            return qFromLittleEndian<quint16>(ptr)/(float)std::numeric_limits<qint16>::max() - 1;
+        else
+            return qFromBigEndian<quint16>(ptr)/(float)std::numeric_limits<qint16>::max() - 1;
+    } else if (format.sampleSize() == 16 && format.sampleType() == QAudioFormat::SignedInt) {
+        if (format.byteOrder() == QAudioFormat::LittleEndian)
+            return qFromLittleEndian<qint16>(ptr)/(float)std::numeric_limits<qint16>::max();
+        else
+            return qFromBigEndian<qint16>(ptr)/(float)std::numeric_limits<qint16>::max();
+    } else if (format.sampleSize() == 32 && format.sampleType() == QAudioFormat::UnSignedInt) {
+        if (format.byteOrder() == QAudioFormat::LittleEndian)
+            return qFromLittleEndian<quint16>(ptr)/(float)std::numeric_limits<qint16>::max() - 1;
+        else
+            return qFromBigEndian<quint16>(ptr)/(float)std::numeric_limits<qint16>::max() - 1;
+    } else if (format.sampleSize() == 32 && format.sampleType() == QAudioFormat::SignedInt) {
+        if (format.byteOrder() == QAudioFormat::LittleEndian)
+            return qFromLittleEndian<qint32>(ptr)/(float)std::numeric_limits<qint32>::max();
+        else
+            return qFromBigEndian<qint32>(ptr)/(float)std::numeric_limits<qint32>::max();
+    } else if (format.sampleSize() == 32 && format.sampleType() == QAudioFormat::Float) {
+        return *reinterpret_cast<const float *>(ptr);
+    }
+
+    return 0.;
+}
+#elif QT_VERSION_MAJOR == 6
+#include <QMediaDevices>
+#include <QAudioSource>
+#include <QAudioSink>
+#include <QAudioDevice>
+
+float normalizeSampleValue(const QAudioFormat& format, const void *sample){
+    return format.normalizedSampleValue(sample);
+}
+
+#endif
+
+namespace PokemonAutomation{
+
+// QIODevice defines an interface for IO.
+// QAudioSource reads audio from a device and writes data to QIODevice,
+// while QAudioSink loads data from QIODevice and sends to a device.
+class AudioIODevice : public QIODevice
+{
+    // Q_OBJECT
+
+public:
+    AudioIODevice(AudioDisplayWidget* parent);
+
+    qint64 writeData(const char *data, qint64 len) override;
+    
+    qint64 readData(char *data, qint64 maxlen) override;
+    qint64 bytesAvailable() const override;
+    qint64 size() const override;
+    bool isSequential() const override { return true; }
+
+private:
+    void computeFFT();
+
+private:
+    AudioDisplayWidget* m_parent = nullptr;
+    // Large buffer to store incoming audio signals.
+    std::vector<float> m_audioFloatBuffer;
+    // The index on m_audioFloatBuffer to store next incoming audio signal.
+    size_t m_floatBufferNextWrite = 0;
+
+    // The buffer to store fft input
+    std::vector<float> m_fftInputBuffer;

```

Comment:
These will need to be aligned. Use the AlignedVector class. You're probably getting lucky it isn't crashing since Unix systems align to 16 bytes by default and right now the default mode is SSE4.2 which is also only 16 bytes.